### DISCUSSION TITLE
ObjectiveSQL vs MyBatis

### DISCUSSION THREAD

**Discussion by catchex** [braisdom/ObjectiveSql#78] [General]
The ObjectiveSQL is a new ORM framework, it enhances the existed ORM framework capability, the first is simple querying.

### Definition of domain model

```java
@DomainModel
public class Blog {
    private Long id;
    private String title;
    private Integer state;

    @Relation(relationType = RelationType.BELONGS_TO)
    private Author author;
}
```

In MyBatis:

```xml
<mapper>
  <resultMap type="Blog" id="blog">
    <id column="id" property="id" />
    ...
    <association property="author" javaType="Author">
      ...
    </association>
  </resultMap>
  <select id="findBlog" parameterType="int" resultMap="blog">
    ...
  </select>
</mapper>
```

```java
...
BlogMapper mapper = session.getMapper(BlogMapper.class); 
Blog blog = mapper. findBlog(...);
```

In ObjectiveSQL:

```java
Blog blog = Blog.queryByPrimaryKey(1, Blog.BELONGS_TO_AUTHOR);
```

**Comment by myangyu** [comment]
ObjectiveSQL 是不是应该和Hibernate对比呢？他们都是摒弃了直接拼装sql。和mybatis是不是不太对标？

**Comment by catchex** [comment]
@myangyu 不管是Hibernate 还是MyBatis 还是其它ORM 框架，最终都是在拼装SQL，只是提供的API 不一样而已。

**Reply by myangyu** [reply]
所以如果和Hibernate对比的话，它有那些优势呢？

**Reply by catchex** [reply]
1）针对简单SQL：ObjectiveSQL 的自动代码生成 和Hibernate 比较类似，但从API 上的使用上更易读。
2）针对复杂SQL：Hibernate 没有很好的支持，只能通过字符串的形式写，但ObjectiveSQL 提供易读、灵活的API，你可以仔细看一下复杂SQL 那部分

**Reply by myangyu** [reply]
好的，感谢。项目中再尝试用轻量级的orm规避与db的交互，专注于业务。ObjectiveSQL看上去是非常不错的选择👍。

**Reply by catchex** [reply]
我也是写了很多年复杂SQL 后，才会写一个工具能够避免直接写SQL的，最近我在做一个智能SQL 生成的工具，只需要让程序员写Select 和Where 即可，不用关心Join，Subquery 等概念，写好了通知你。
